### PR TITLE
Implement support to `MessageEntityMentionName`

### PR DESCRIPTION
**Pull request by AndrielFR**
With this, you can mention users either using html or markdown.

### CODE REVIEW

File: lib/grammers-client/src/parsers.rs [hunk 1]
```diff
@@ -117,12 +117,21 @@ pub fn parse_markdown_message(message: &str) -> (String, Vec<tl::enums::MessageE
             update_entity_len!(Italic(offset) => entities);
         }
 
-        // [text link](https://example.com)
+        // [text link](https://example.com) or [user mention](tg://user?id=12345678)
         Event::Start(Tag::Link(_kind, url, _title)) => {
-            push_entity!(MessageEntityTextUrl(offset, url = url.to_string()) => entities);
+            if url.starts_with("tg://user?id") {
```
Comment (Lonami) [inline lib/grammers-client/src/parsers.rs:122]: Please refactor this into a `const MENTION_URL_PREFIX` (name can be better).

File: lib/grammers-client/src/parsers.rs [hunk 2]
```diff
@@ -117,12 +117,21 @@ pub fn parse_markdown_message(message: &str) -> (String, Vec<tl::enums::MessageE
...
+                let user_id = url[13..].parse::<i64>().unwrap();
```
Comment (Lonami) [inline lib/grammers-client/src/parsers.rs:123]: Once we have the const, use its `.len()`.

File: lib/grammers-client/src/parsers.rs [hunk 3]
```diff
@@ -117,12 +117,21 @@ pub fn parse_markdown_message(message: &str) -> (String, Vec<tl::enums::MessageE
...
+                push_entity!(MessageEntityMentionName(offset, user_id = user_id) => entities);
+            } else {
+                push_entity!(MessageEntityTextUrl(offset, url = url.to_string()) => entities);
+            }
         }
-        Event::End(Tag::Link(_kindd, _url, _title)) => {
-            update_entity_len!(TextUrl(offset) => entities);
+        Event::End(Tag::Link(_kindd, url, _title)) => {
+            if url.starts_with("tg://user?id") {
```
Comment (Lonami) [inline lib/grammers-client/src/parsers.rs:130]: Same here.

File: lib/grammers-client/src/parsers.rs [hunk 4]
```diff
@@ -237,8 +246,14 @@ pub fn parse_html_message(message: &str) -> (String, Vec<tl::enums::MessageEntit
                             .map(|a| a.value.to_string())
                             .unwrap_or_else(|| "".to_string());
 
-                        push_entity!(MessageEntityTextUrl(self.offset, url = url)
-                            => self.entities);
+                        if url.starts_with("tg://user?id") {
```
Comment (Lonami) [inline lib/grammers-client/src/parsers.rs:249]: Same `const`.

File: lib/grammers-client/src/parsers.rs [hunk 5]
```diff
@@ -237,8 +246,14 @@ pub fn parse_html_message(message: &str) -> (String, Vec<tl::enums::MessageEntit
...
+                            let user_id = url[13..].parse::<i64>().unwrap();
```
Comment (Lonami) [inline lib/grammers-client/src/parsers.rs:250]: `const`'s `.len()`.

File: lib/grammers-client/src/parsers.rs (hunk 1 above)
Comment (AndrielFR) [reply lib/grammers-client/src/parsers.rs:122]: ok.

**Review by Lonami — COMMENTED** [review]
The current CI error is unrelated.

The next step is to change `Client::send_message` to look for the message entities, replacing all these `MessageEntityMentionName` with [`InputMessageEntityMentionName`](https://docs.rs/grammers-tl-types/latest/grammers_tl_types/enums/enum.MessageEntity.html#variant.InputMessageEntityMentionName).

**Comment by Lonami** [comment]
Just to confirm, if you use this code to send a message with a mention, the mention is removed by the server, right? I don't think things have changed since I tried, but maybe.

**Comment by AndrielFR** [comment]
> Just to confirm, if you use this code to send a message with a mention, the mention is removed by the server, right? I don't think things have changed since I tried, but maybe.

yes, i can change all to `InputMessageEntityMention` to avoid it?

**Comment by Lonami** [comment]
Yes that's the next step (I detailed it some more in the Telegram group).

**Comment by AndrielFR** [comment]
> Yes that's the next step (I detailed it some more in the Telegram group).

i'm doing the function to parse it.

**Review by Lonami — CHANGES_REQUESTED** [review]
We can merge this once we document this special link in https://github.com/Lonami/grammers/blob/4e95ca5e37fc6019b06b73b9d2dade9b8561e9cd/lib/grammers-client/src/types/input_message.rs#L313 (and the `html`). Something like "inline links with this format (…) will be replaced with inline mentions when possible".

**Comment by AndrielFR** [comment]
> We can merge this once we document this special link in
> 
> https://github.com/Lonami/grammers/blob/4e95ca5e37fc6019b06b73b9d2dade9b8561e9cd/lib/grammers-client/src/types/input_message.rs#L313
> (and the `html`). Something like "inline links with this format (…) will be replaced with inline mentions when possible".

the way i did is ok?

**Comment by Lonami** [comment]
Yes, docs are fine, but `rustfmt` is complaining.

**Comment by AndrielFR** [comment]
> Yes, docs are fine, but `rustfmt` is complaining.

done.

**Comment by Lonami** [comment]
Thank you!

**Pull request merged**